### PR TITLE
Expose authorizer to ValidatingAdmissionPolicy message expressions.

### PR DESCRIPTION
**Pull request by benluddy**
#### What type of PR is this?

/kind feature
/kind api-change

#### What this PR does / why we need it:

Binding `authorizer` for `messageExpression` makes the evaluation environment identical across ValidatingAdmissionPolicy's `expression` and `messageExpression` fields. It also enables policy authors to include secondary authz decision reasons within error messages:

```
spec:
  validations:
  - expression: "authorizer.requestResource.check('foo').allowed()"
    messageExpression: "'not allowed to foo: ' + authorizer.requestResource.check('foo').reason()"
```

Cost concerns related to identical authorization checks appearing in both `expression` and `messageExpression` prompted the decision not to bind `authorizer` to the `messageExpression` evaluation environment. Those concerns may be addressed with https://github.com/kubernetes/kubernetes/pull/116443.

#### Does this PR introduce a user-facing change?

```release-note
ValidatingAdmissionPolicy message expressions may now access the "authorizer" variable that had previously been accessible only to validation expressions.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
```

### CODE REVIEW

**Comment by irvifa** [comment]
Do we have an issue opened for the PR?

**Review by irvifa — APPROVED** [review]

File: pkg/apis/admissionregistration/validation/validation_test.go
Comment (alexzielenski) [inline pkg/apis/admissionregistration/validation/validation_test.go:2650]: I believe this is a test only of validation. Could you add tests that uses the authorizer value in a message expression? If its possible to show that the caching is working across the rule expressions and the message expressions that would also be good to see

**Review by alexzielenski — COMMENTED** [review]
Looks good. I think I'd like to see tests of the added functionality

File: pkg/apis/admissionregistration/validation/validation_test.go
Comment (benluddy) [reply pkg/apis/admissionregistration/validation/validation_test.go:2650]: Thanks for catching this. I mistakenly thought I had included integration test coverage too. Added a new test case in my last push. I'll add an expression/messageExpression caching test once the caching PR merges.

**Comment by benluddy** [comment]
/assign @liggitt

Please take a look. I've removed the caveat about authorizer being unavailable to message expressions.

**Comment by liggitt** [comment]
also, is the cost estimation aware of the authz cache, so it can account for additional calls properly? if not, what budget are we giving the messageExpression and will calling the authorizer even once exceed that budget?

**Comment by liggitt** [comment]
Are we sure the authorization reason is useful enough and consistent enough to surface in admission messages this way? most places we do secondary authz checks, we don't return the reason to the user when unauthorized:

https://github.com/kubernetes/kubernetes/blob/4af23c157c3416dbbbd34c94204959c564322a59/plugin/pkg/admission/certificates/util.go#L34-L57

https://github.com/kubernetes/kubernetes/blob/4af23c157c3416dbbbd34c94204959c564322a59/pkg/registry/admissionregistration/validatingadmissionpolicybinding/authz.go#L102-L108

https://github.com/kubernetes/kubernetes/blob/4af23c157c3416dbbbd34c94204959c564322a59/pkg/registry/rbac/escalation_check.go#L90-L97

**Comment by benluddy** [comment]
> do we guarantee use of `authorizer` in `messageExpression` will hit the cache?

No, only that repeated identical checks in the course of admitting a given request will hit the cache. If you were to make a different check within a messageExpression that doesn't appear elsewhere, that would generate a real authorize call.

> also, is the cost estimation aware of the authz cache, so it can account for additional calls properly? if not, what budget are we giving the messageExpression and will calling the authorizer even once exceed that budget?

The cost accounting assumes the full cost every time. MessageExpression evaluations have the same per-evaluation budget as validation expressions and permit at most two authz checks.

> Are we sure the authorization reason is useful enough and consistent enough to surface in admission messages this way

I'm not sure. We do already expose it to validation expressions, so at least in theory, a reason may currently be the difference between admit and deny. If it's not useful enough to surface in messages, then maybe that was a mistake?

The decision itself does seem useful for varying error messages between secondary authorization denials and other failures within an expression, and I think there is usability value in making the evaluation environment consistent across validation and message expressions.

**Comment by cici37** [comment]
Extra context:
1. We would expect the same authorizer check appear in both validation expr and messageExpression so only the repeated identical checks hit the cache might be good enough in this case. And the [variable composition](https://github.com/kubernetes/kubernetes/pull/118642) might also help in this case. 
2. For cost, even the same per-evaluation budget will apply to messageExpression but it will share a same overall budget with validation expr which is 10 times of the per-evaluation budget. 
3. Currently the reason would only represent the validation expr failure since we would ignore/log the messageExpression failure as long as it's not critical.

**Comment by liggitt** [comment]
> The decision itself does seem useful for varying error messages between secondary authorization denials and other failures within an expression

Wouldn't that be better modeled by having a single rule that does the authorization check with an associated message related to authorization, then having other rules doing other checks with other associated messages?

> I think there is usability value in making the evaluation environment consistent across validation and message expressions.

In the abstract, yes, but I'm having a hard time thinking of an actual use of `authorizer` in messageExpression I would be happy with or recommend.

**Comment by benluddy** [comment]
> Wouldn't that be better modeled by having a single rule that does the authorization check with an associated message related to authorization, then having other rules doing other checks with other associated messages?

Does that support a use case where secondary authorization success bypasses a more restrictive subexpression, e.g. `authorizer.requestResource.check('foo') ? true : object.spec.bar == 'baz'`?

**Comment by liggitt** [comment]
> Does that support a use case where secondary authorization success bypasses a more restrictive subexpression, e.g. `authorizer.requestResource.check('foo') ? true : object.spec.bar == 'baz'`?

Hmm... good point... probably not.

Would the variable composition work be a better way to address this, where the authorizer use / check could be run a single time, the result stored and referenced from the rule or messageExpression?

**Comment by valaparthvi** [comment]
Hey @benluddy 
On behalf of the bug triage release team for 1.28, I'd like to inform you about the Code Freeze ([01:00 UTC Wednesday 19th July 2023 / 18:00 PDT Tuesday 18th July 2023](https://everytimezone.com/s/72ee8496)) that approaches tomorrow. Please ensure this PR goes in by tomorrow to avoid spill over to the next release.

cc: @kubernetes/sig-api-machinery-leads

**Comment by Atharva-Shinde** [comment]
[exception request received](https://groups.google.com/a/kubernetes.io/g/release-team/c/NtaSED_F1a8/m/2JYiKx2MAwAJ).
Holding for it's approval.
/hold

**Comment by cici37** [comment]
The exception has been approved.
/unhold

@benluddy Would you mind adding tests to verify the cost concern has been addressed? (authorizer properly cached and not increase the cost if specified inside messageExpression?) Thank you

**Comment by cici37** [comment]
Update: Due to the offline discussion, it might make more sense to have user access Authz check from messageExpression through variables only since there is no obvious value of exposing authz lib in messageExpression directly except for the consideration of unifying access, and we wanna encourage user to not adding authz check in messageExpression under consideration of cost and impact on number of validation allowed. We agreed to drop this feature for 1.28.
/remove-milestone v1.28

**Comment by liggitt** [comment]
I see three possible scenarios where `authorizer` is used in `messageExpression`:

1. the user does an identical check in `messageExpression` to something in `rule`, we cache it so they don't pay the cost, but we would rather they have used a variable and explicitly reused the result

2. the user _meant_ to do an identical check in `messageExpression` and `rule`, but typoed it somehow and did something slightly different, they pay the cost of an additional authz check, and their message is wrong

3. the user meant to do a _different_ check in `messageExpression` than in `rule`... they pay the cost of an additional authz check, and I have a difficult time imagining a case where a novel authz check solely for use in `messageExpression` would be useful

Since variables are available, I think we should push them to use those when they want to share an authz result between `rule` and `messageExpression`, and if there are legitimate cases for 3 that we're missing, use of a variable enables those as well.

**Comment by liggitt** [comment]
dropping from api-review queue per https://github.com/kubernetes/kubernetes/pull/118819#issuecomment-1643083008

/remove-label api-review

**Pull request closed without merging**
